Sources/Prelude/Refinements.swift

```diff
@@ -0,0 +1,136 @@
+//
+// This source file is part of Prelude, an open source project by Wayfair
+//
+// Copyright (c) 2018 Wayfair, LLC.
+// Licensed under the 2-Clause BSD License
+//
+// See LICENSE.md for license information
+//
+
+/// protocol that represents a rule that can be used to “refine” values of type `RefinedType`. Implement the `static Refinement.isValid(_:)` function to describe which values are allowed.
+public protocol Refinement {
+    /// the type being “refined”
+    associatedtype RefinedType
+
+    /// implement this function to describe which values of type `RefinedType` should be allowed
+    ///
+    /// - Parameter value: a value of type `RefinedType`
+    /// - Returns: `true` if this rule should allow `value`
+    static func isValid(_ value: RefinedType) -> Bool
+}
+
+/// error that is thrown when a value fails a refinement. Examine `localizedDescription` to determine the details of the failure
+public struct RefinementError: Error {
+    public let localizedDescription: String
+
+    init(_ localizedDescription: String) {
+        self.localizedDescription = localizedDescription
+    }
+}
+
+/// a container for a “refined” type. The type parameter `A` represents the underlying value (eg. `Int`). The type parameter `R` represents the “refinement” (eg. `NonZero`, `NotNegative`, `GreaterThanFifty`, etc.)
+public struct Refined<A, R: Refinement> where A == R.RefinedType {
+    /// the underlying refined value. This value is guaranteed to satisfy the refinement represented by `R`
```

Comment:
… unless it’s a mutable reference type, in which case all bets are off. Trying to decide if I need to write that or not 😆